### PR TITLE
Eng Common Sync Improvements.

### PR DESCRIPTION
**Pull request by chidozieononiwu**
- Refactor GitHub API call to allow reusability.
- Add Clean up pipeline with step that deletes upstream sync branches . Moved to https://github.com/Azure/azure-sdk-tools/pull/1088
- Add auto-merge label to Sync and Tools PRs
- Remove the Verify and Merge stage.

### CODE REVIEW

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:21]: We should file a tracking issue about the race condition as we should eliminate that problem.

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:22]: "till" -> "until"

Comment (weshaggard) [reply eng/common/README.md:22]: We should only do the release of the template package if the changes might impact the release pipeline.

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:23]: Do we think it is safe to use "auto-merge" here? With your new patch file approach it might be but with the older approach I would say it is not safe enough to auto-merge.

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:24]: Just to clearify this step I'd suggest saying something like "Go review and approve each of your Sync PRs. The merging will happen automatically."

File: eng/pipelines/eng-common-sync.yml
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:150]: What blocks someone from merging their Tools PR before all the Sync PR's are merged?

File: eng/pipelines/eng-common-sync.yml
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:133]: Won't this cause the tools PR to get merged immediately after this step/pipeline finishes and thus before the sync PRs?

File: eng/common/scripts/Add-Issue-Comment.ps1
Comment (weshaggard) [inline eng/common/scripts/Add-Issue-Comment.ps1:28]: Why are these reading from environment variables as opposed to the passed in variables? I would not have guess that this parameter the name of an environment variable.

Comment (weshaggard) [reply eng/common/scripts/Add-Issue-Comment.ps1:28]: If someone wants to pass in an environment variable they should let the caller do that.

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHub-API.ps1:5]: @mitchdenny do you think using secure string and passing separate parameters is easier or better then just passing the headers directly?

File: eng/common/scripts/Add-Issue-Comment.ps1
Comment (weshaggard) [inline eng/common/scripts/Add-Issue-Comment.ps1:42]: If we are going to provide helpers for invoking github apis we should include the error handling there as well

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHub-API.ps1:2]: Given you are trying to provide a library of functions we shouldn't pass in the AuthToken at this level it should be passed in as a parameter to the individual methods. We need to be able to pass different tokens for different calls and we also should try to avoid having this being in this scope.

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHub-API.ps1:7]: $token should also be a parameter to these methods.

File: eng/common/scripts/Add-Issue-Comment.ps1
Comment (weshaggard) [inline eng/common/scripts/Add-Issue-Comment.ps1:26]: We should move this into common.ps1 and import common.ps1 here instead of this and logging independently.

File: eng/common/scripts/Delete-Remote-Branches.ps1
Comment (weshaggard) [inline eng/common/scripts/Delete-Remote-Branches.ps1:9]: We should be importing common.ps1.

File: eng/pipelines/eng-common-sync.yml
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:131]: Why aren't creating a wrapper script for this? It will definitely make diagnosing issues easier.

File: eng/common/scripts/Delete-Remote-Branches.ps1
Comment (weshaggard) [inline eng/common/scripts/Delete-Remote-Branches.ps1:15]:
```suggestion
$syncBranches = git branch -r --list origin/$BranchPrefix* | % { $_ -replace "origin/", "" }
```

File: eng/common/scripts/Delete-Remote-Branches.ps1
Comment (weshaggard) [inline eng/common/scripts/Delete-Remote-Branches.ps1:28]: We need to also have a plan for orphaned pull requests where the tools PR has been closed or merged.

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHub-API.ps1:66]: nit: I'd put a space before the `{` on these lines.

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHub-API.ps1:137]: Doesn't split members already return a array? Do we need to wrap it in another array?

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHub-API.ps1:176]: Does this need to be wrapped in an array?

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [reply eng/common/scripts/Invoke-GitHub-API.ps1:137]: Should we also verify that the assignees is not null/empty? Or do we want to support case as I suspect it will remove the assignees.

File: eng/pipelines/eng-common-sync.yml
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:3]: I assume we don't need the PRDataFileName any more either.

File: eng/pipelines/templates/steps/sync-directory.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:5]: Why change the name here?

File: eng/common/scripts/Add-Issue-Comment.ps1
Comment (weshaggard) [reply eng/common/scripts/Add-Issue-Comment.ps1:28]: I'm also not sure there is much value in splitting this into 3 parameters and instead suggest just one parameter for the comment and let the caller concat if they like.

File: eng/common/README.md
Comment (chidozieononiwu) [reply eng/common/README.md:21]: https://github.com/Azure/azure-sdk-tools/issues/1087

File: eng/common/README.md
Comment (chidozieononiwu) [reply eng/common/README.md:23]: We can have the `auto-merge` happen only when the pipeline is triggered via a pull request. i.e. only when it uses the patch files.

File: eng/common/scripts/Add-Issue-Comment.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/Add-Issue-Comment.ps1:42]: Am not sure how to properly do that. Should I use the try catch inside the `Invoke-GitHub-API.ps1` script. if so should I also use the `logging.ps1` to log the errors there? Won't that cause DevOps to exit prematurely?

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/Invoke-GitHub-API.ps1:137]: I have `[ValidateNotNullOrEmpty()]` to take care of that.

File: eng/pipelines/eng-common-sync.yml
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml:150]: I'll add back this stage.

File: eng/pipelines/templates/steps/sync-directory.yml
Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/sync-directory.yml:5]: To better reflect whats going on since we no longer create the pull request using that parameter.

File: eng/common/README.md
Comment (mitchdenny) [inline eng/common/README.md:21]: If this does happen, what is the recovery step, delete the sync PRs and trigger again?

File: eng/common/scripts/Add-Issue-Labels.ps1
Comment (mitchdenny) [inline eng/common/scripts/Add-Issue-Labels.ps1:22]: nit: Why isn't this and the filename following the Verb-Noun convention?

File: eng/common/scripts/Delete-Remote-Branches.ps1
Comment (mitchdenny) [inline eng/common/scripts/Delete-Remote-Branches.ps1:1]: nit: Filename should be Verb-Noun.

File: eng/common/scripts/Delete-Remote-Branches.ps1
Comment (mitchdenny) [inline eng/common/scripts/Delete-Remote-Branches.ps1:23]: nit: Verb-Noun

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (mitchdenny) [inline eng/common/scripts/Invoke-GitHub-API.ps1:7]: For auth we should just use the -Token parameter and -Authentication parameter.

File: eng/pipelines/templates/steps/sync-directory.yml
Comment (mitchdenny) [inline eng/pipelines/templates/steps/sync-directory.yml:5]: Better capitalization would be UpstreamBranchName

**Review by mitchdenny — CHANGES_REQUESTED** [review]

File: eng/common/README.md
Comment (chidozieononiwu) [reply eng/common/README.md:21]: There is no need to delete the sync pipeline. You'll just need to rerun the particular template pipeline that experienced the race condition. Again you only have to do this if your changes will potentially affect release.

File: eng/common/scripts/Add-Issue-Labels.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/Add-Issue-Labels.ps1:22]: You mean it should be `Add-IssueLabels` ?

File: eng/pipelines/eng-common-sync.yml
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml]: Ok I see what you mean. I switched back to using the verify and merge.

File: eng/pipelines/templates/jobs/tools-cleanup.yml
Comment (chidozieononiwu) [reply eng/pipelines/templates/jobs/tools-cleanup.yml]: PR Created https://github.com/Azure/azure-sdk-tools/pull/1088/files

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/Invoke-GitHub-API.ps1]: @mitchdenny

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/Invoke-GitHub-API.ps1:7]: Wes had a question about that here https://github.com/Azure/azure-sdk-tools/pull/1081#discussion_r502647956

File: eng/common/scripts/Add-Issue-Comment.ps1
Comment (weshaggard) [reply eng/common/scripts/Add-Issue-Comment.ps1]: That is a fair point perhaps we should let the error just flow and if someone wishes to handle it then can do it correctly in their context.

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (weshaggard) [reply eng/common/scripts/Invoke-GitHub-API.ps1]: Isn't setting it to empty a potentially valid thing to do if you want to clear the assignee?

Comment (chidozieononiwu) [reply eng/common/scripts/Invoke-GitHub-API.ps1]: Not for this method. To o that one can use the `Update-Issue` method. Which is an `PATCH` rather than `POST` method.

File: eng/pipelines/eng-common-sync.yml
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:179]: What value to you think we have by adding the auto-merge label here? Is it only to prevent someone from needing to click the merge button themselves?

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:25]:
```suggestion
7. Sign Off on the VerifyAndMerge stage. This will merge any remaining open **Sync PR** and also append `auto-merge` to the **Tools PR**.
```

File: eng/common/scripts/Invoke-GitHubAPI.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:3]: I suggest calling this `Get-GithubHeaders` since it is specific to our github usage.

File: eng/pipelines/templates/steps/sync-directory.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:86]: Don't you need to update the name of the script here?

File: eng/common/scripts/Invoke-GitHubAPI.ps1
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:165]: Isn't it valid to pass in empty to clear the assignees?

File: eng/pipelines/eng-common-sync.yml
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml:179]: We gain some speed in coordinating the merge. If the Sync PRs are merged then we want to merge the Tools PR and soon as is possible. The auto-merge will help in the case the user forgets and the PR is just waiting.

File: eng/common/scripts/Invoke-GitHubAPI.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/Invoke-GitHubAPI.ps1:165]: Not for this function. That can be done with the Update-Issue function.

Comment (weshaggard) [reply eng/common/scripts/Invoke-GitHubAPI.ps1:165]: Not that I think it is important but it would appear as though the Update-Issue function also blocks empty.

File: eng/pipelines/templates/steps/sync-directory.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:89]: It is probably worth adding a link to the workflow in this common as well so that anyone new that submits a PR to the tools repo will have a link to the docs for the workflow.

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:18]:
```suggestion
1. Create a PR (**Tools PR**) in the `azure-sdk-tools` repo with changes to eng/common directory.
```

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:20]:
```suggestion
3. The  `azure-sdk-tools - sync - eng-common` pipeline queues test runs for template pipelines in various languages. These help to you test your changes in the **Tools PR**.
```

File: eng/common/scripts/Invoke-GitHubAPI.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/Invoke-GitHubAPI.ps1:165]: Oh, I see what you mean. I should check for whitespace in this.
For `Update-Issue` you can pass white space " " to clear labels or assignees.

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:22]:
```suggestion
5. If you make additional changes to your **Tools PR** repeat steps 1 - 4 until you have completed the necessary testing of your changes. This includes full releases of the template package, if necessary. 
```

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:23]:
```suggestion
6. Sign off on CreateSyncPRs stage of the sync pipeline using the approval gate. This stage will create the **Sync PRs** in the various language repos with the `auto-merge` label applied. A link to each of the **Sync PRs** will show up in the **Tools PR** for you to click and review.
```

File: eng/common/README.md
Comment (weshaggard) [inline eng/common/README.md:25]:
```suggestion
7. Sign off on VerifyAndMerge stage of the sync pipeline using the approval gate. This stage will merge any remaining open **Sync PRs** and also append `auto-merge` to the **Tools PR** so it will automatically merge once the pipeline finishes.
```

**Review by weshaggard — APPROVED** [review]
I added some text changes to the README but otherwise the changes look good to me. Thanks for adding the github api helpers and improving the process.

File: eng/common/scripts/Invoke-GitHub-API.ps1
Comment (mitchdenny) [reply eng/common/scripts/Invoke-GitHub-API.ps1]: I'm not that hung up on it, but I think given that the Cmdlet has support for exactly what we want to do we should use it. I'm not a fan of secure string either since in our use case its not secure.

**Pull request merged**
